Back and forward substitution

ESeNonFossiIo opened the pull request.

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -662,6 +662,177 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
+ * @relates LinearOperator
+ *
+ * This function implement a gauss elimination argument to invert a lower
+ * block triangular matrix.
```
I believe this process is mostly known as Backward substitution (or Forward), i.e.,

https://en.wikipedia.org/wiki/Triangular_matrix#Forward_and_back_substitution

May be rename it and insert a Mathematical description of the algorithm as in the link above, adapted for matrices?

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -662,6 +662,204 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
+ * ...        ...
+ * A0n x0 + A1n x1 + ... + Ann xn = yn
+ *
+ * First fo all, x0 = A00^-1 y0
```
typo: First *of* all

bangerth commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -662,6 +662,204 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
+template <unsigned int n, typename Range, typename Domain, typename BlockMatrix>
+LinearOperator<Range, Domain>
+block_forward_substitution(const BlockMatrix &block_matrix,
+                           const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, n> &inverse_diagonal)
```
In a call to this declaration, only the template arguments `BlockMatrix` and `n` can be deduced by the compiler, whereas `Range` and `Domain` need to be specified by hand. It may be useful to have these two as the first two template parameters so that only they need to be specified and the trailing template parameters can be deduced automatically.

I imagine `BlockMatrix` can be a linear operator itself, or really anything. Can we ask them for their own `RangeType` and `DomainType`? If so, we could get rid of the `Range` and `Domain` template arguments.

bangerth commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -662,6 +662,204 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
+template <unsigned int n, typename Range, typename Domain, typename BlockMatrix>
+LinearOperator<Range, Domain>
+block_back_substitution(const BlockMatrix &block_matrix,
+                        const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, n> &inverse_diagonal)
```
same here, obviously.

ESeNonFossiIo commented:
I made an attempt to see how it looks like.
Please, can you tell me if it is what you mean?
Should I remove n and Range from template or I can leave them in order to save readability?

bangerth commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -662,6 +662,208 @@ block_triangular_inverse(const BlockMatrix &block_matrix,
+          typename Range  = typename BlockMatrix::BlockType>
+LinearOperator<Range, Range>
+block_forward_substitution(const BlockMatrix &block_matrix,
+                           const std::array<LinearOperator<typename Range::BlockType, typename Range::BlockType>, n> &inverse_diagonal)
```
Does this compile? In other words, does BlockMatrix always have a local static const variable `n` and a local typedef `BlockType`? If so, then yes this works.

I had in mind to write it as
```
template <typename Range, typename Domain, typename BlockMatrix, int n>
```
The last two arguments can be deduced by the compiler, so you don't have to explicitly specify them in the call. The first two cannot be deduced, so you have to explicitly specify them.

luca-heltai commented:
Also for this functions, it should be possible to write them in terms of an array of operators, as you did in the past for the the other two (which, if I understand correctly, should be superseded by this? If this is the case, I'd just remove the other two...)

ESeNonFossiIo commented:
Ok. I have removed all obsolete functions upper/lower_tringular. I have replaced them with new ones.
They use back and forward substitution to emulate the action of an inverse block matrix.
Therefore, I deleted all tests that require upper/lower triangular matrix and i have fixed the test for inverse matrix in order to test new functions.

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,104 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
+ * This function implement a forward substitution argument to invert a lower
+ * block triangular matrix.
+ * It takes as argement an array of array of LinearOperators @p block_matrix
+ * representing a block matrix and an array of LinearOperators
```
a block *triangular* matrix

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,104 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
+ * and therefore:
+ *    xn = Ann^-1 ( yn - A0n x0 - ... - A(n-1)n x(n-1) )
+ *
+ * Caveat: Tvmult and Tvmult_add have not been implemented, yet. This may lead to mistakes.
```
I'd remove the phrase "This may lead to mistake", since the default code of LinearOperator is to throw an exception if called when you do not implement Tvmult...

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -425,213 +431,140 @@ lower_triangular_operator(const std::array<std::array<LinearOperator<typename Ra
+ * This function implement a back substitution argument to invert an upper
+ * block triangular matrix.
+ * It takes as argement an array of array of LinearOperators @p block_matrix
+ * representing a block matrix and an array of LinearOperators
```
representing an *upper triangular block matrix*

luca-heltai commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -425,213 +431,140 @@ lower_triangular_operator(const std::array<std::array<LinearOperator<typename Ra
+ * and therefore:
+ *    x0 = A00^-1 ( y0 - A0n xn - ... - A01 x1 )
  *
+ * Caveat: Tvmult and Tvmult_add have not been implemented, yet. This may lead to mistakes.
```
Same as above. Here and above I'd also specify what is the behavior of this class when you feed it with an array of operators which represents a full block matrix (i.e., this class ignores all lower blocks of the matrix, and above, this class ignores all upper blocks of the matrix).

luca-heltai commented:
To me this looks good. If @tamiko (and @bangerth ) are ok with it, squash all commits and we'll merge.

ESeNonFossiIo commented:
I'd not remove caveat: at first glance it is not clear the behaviour of a linearOperator not having Tvmult implemented.
When I wrote the test i did not know it and I found a segmentation fault. I took me some time to find the error.

bangerth commented:
I will let @tamiko comment as in the past.

tamiko commented:
I'll be back on Monday. Sorry for the delay.

ESeNonFossiIo commented:
ping @luca-heltai
ping @tamiko

tamiko commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,105 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
- * A function that takes an array of arrays of LinearOperators @p block_matrix
- * and returns its associated lower triangular matrix operator
- * (diagonal is not included).
+ * This function implement a forward substitution argument to invert a lower
```
* implements

tamiko commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,105 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
- * (diagonal is not included).
+ * This function implement a forward substitution argument to invert a lower
+ * block triangular matrix.
+ * It takes as argement an array of array of LinearOperators @p block_matrix
```
* "As arguments, it takes an array of arrays of ..."

tamiko commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,105 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
- *  a20 | a21 | a22             a20 | a21 |
- * @endcode
+ * Let us assume we have a linear system where each coefficient of the system is a
+ * matrix:
```
* Let us assume we have a linear system with the following block structure:

tamiko commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -325,98 +325,105 @@ block_diagonal_operator(const LinearOperator<typename Range::BlockType, typename
+ *    xn = Ann^-1 ( yn - A0n x0 - ... - A(n-1)n x(n-1) )
+ *
+ * Notice that we are not using the whole matrix: just the lower triangular block
+ * matrix obtained from @p block_matrix it is used.
```
Can you change this into a <code>@note</code>?

tamiko commented on include/deal.II/lac/block_linear_operator.h:
```diff
@@ -425,213 +432,141 @@ lower_triangular_operator(const std::array<std::array<LinearOperator<typename Ra
  * @relates LinearOperator
  *
- * This function is a specification of the above function that
- * allows to work with block matrices @p block_matrix .
```
Same minor typos in the second function :-)

tamiko commented:
Just those minor typos and we can merge.

tamiko commented:
@luca-heltai I think we should discuss some redesign/general improvements for the <code>BlockLinearOperator</code> stuff next week. I have some ideas, but - unfortunately - not much time in the last weeks to prototype those.
